src/datetime.ts

```diff
@@ -232,55 +231,91 @@ export function dateTimeExpr(d: DateTime | DateTimeExpr, normalize = false, toJS
   }
 
   if (d.year !== undefined) {
-    units.push(d.year);
+    parts.push(d.year);
   } else if (d.day !== undefined) {
     // Set year to 2006 for working with day since January 1 2006 is a Sunday
-    units.push(SUNDAY_YEAR);
+    parts.push(SUNDAY_YEAR);
   } else {
-    units.push(0);
+    parts.push(0);
   }
 
   if (d.month !== undefined) {
     const month = normalize ? normalizeMonth(d.month) : d.month;
-    units.push(month);
+    parts.push(month);
   } else if (d.quarter !== undefined) {
     const quarter = normalize ? normalizeQuarter(d.quarter) : d.quarter;
-    units.push(quarter + '*3');
+    parts.push(isNumber(quarter) ? quarter * 3 : quarter + '*3');
   } else {
-    units.push(0); // months start at zero in JS
+    parts.push(0); // months start at zero in JS
   }
 
   if (d.date !== undefined) {
-    units.push(d.date);
+    parts.push(d.date);
   } else if (d.day !== undefined) {
     // HACK: Day only works as a standalone unit
     // This is only correct because we always set year to 2006 for day
     const day = normalize ? normalizeDay(d.day) : d.day;
-    units.push(day + '+1');
+    parts.push(isNumber(day) ? day + 1 : day + '+1');
   } else {
-    units.push(1); // Date starts at 1 in JS
+    parts.push(1); // Date starts at 1 in JS
   }
 
   // Note: can't use TimeUnit enum here as importing it will create
   // circular dependency problem!
   for (const timeUnit of ['hours', 'minutes', 'seconds', 'milliseconds'] as const) {
     const unit = d[timeUnit];
-    units.push(typeof unit === 'undefined' ? 0 : unit);
+    parts.push(typeof unit === 'undefined' ? 0 : unit);
   }
 
-  const unitsString = units.join(', ');
+  return parts;
+}
 
-  if (toJSON) {
-    if (d.utc) {
-      return new Function(`return +new Date(Date.UTC(${unitsString}))`)();
-    } else {
-      return new Function(`return +new Date(${unitsString})`)();
-    }
+/**
+ * Return Vega expression for a date time.
+ *
+ * @param d the date time.
+ * @returns the Vega expression.
+ */
+export function dateTimeToExpr(d: DateTime) {
+  const parts: (string | number)[] = dateTimeParts(d, true);
+
+  const string = parts.join(', ');
+
+  if (d.utc) {
+    return `utc(${string})`;
+  } else {
+    return `datetime(${string})`;
   }
+}
+
+/**
+ * Return Vega expression for a date time expression.
+ *
+ * @param d the internal date time object with expression.
+ * @returns the Vega expression.
+ */
+export function dateTimeExprToExpr(d: DateTimeExpr) {
+  const parts: (string | number)[] = dateTimeParts(d, false);
+
+  const string = parts.join(', ');
+
+  if (d.utc) {
+    return `utc(${string})`;
+  } else {
+    return `datetime(${string})`;
+  }
+}
+
+/**
+ * @param d the date time.
+ * @returns the timestamp.
+ */
+export function dateTimeToTimestamp(d: DateTime) {
+  const parts: (string | number)[] = dateTimeParts(d, true);
 
   if (d.utc) {
-    return `utc(${unitsString})`;
+    return +new Date(Date.UTC(...(parts as [any, any])));
```

Comment:
I want to double check: will this work when `parts` contains some `string` elements?

Reply:
No, it doesn't need to but this shouldn't be an issue since we normalize strings to numbers. Let me check what happens if a user doesn't respect the schema.

Reply:
Looks like js accepts this: `new Date("2019", "3", "2")` so we are good. Thanks for checking.

Reply:
I added some code to always support numbers even if they are in strings.